src/main/java/com/github/codeboy/mcide/ide/gui/ProjectMenu.java

```diff
@@ -43,29 +44,45 @@ protected int getPrevButtonIndex() {
         return 50;
     }
 
+    @Override
+    protected Gui createPage() {
+        Gui gui = super.createPage();
+        ItemStack item = createItem(Material.STAINED_GLASS_PANE, "-");
+        for (int i = 0; i < 9; i++) {
+            gui.addItem(item, 54 - 18 + i, Action.none);
+        }
+        return gui;
+    }
+
     public void addFile(MCCodeFile file, boolean mainFile) {
         ItemStack fileItem = createItem(Material.BOOK_AND_QUILL, file.getName(), mainFile);
-        addItem(fileItem, p -> project.editFile(file, p));
+        ArrayList<Gui> pages = getPages();
+        int index = 0;
+        Gui lastPage;
+        do {
+            lastPage = pages.get(index++);
+        } while (lastPage.getFreeSlot() > filesPerPage && index < pages.size());
+        int freeSlot = lastPage.getFreeSlot();
+        if (freeSlot > filesPerPage) {
+            addNextButton(lastPage);
+            lastPage = createPage();
+            pages.add(lastPage);
+            addPrevButton(lastPage);
+        }
+        lastPage.addItem(fileItem, p -> project.editFile(file, p));
     }
 
     public void removeFile() {
-        ArrayList<Gui>pages=getPages();
-        for (int i = pages.size() - 1; i >= 0; i--) {
-            Gui page=pages.get(i);
-            Inventory inventory= page.getInventory();
-
-            int index = inventory.getSize() - 1;
-            while (index>=0) {
-                if(isProjectFile(inventory.getItem(index))){
-                    inventory.setItem(index,null);
-                    return;
-                }
-                index--;
-            }
+        int files = project.getMCCodeFiles().size();
+        if (!project.getMCCodeFiles().get(files - 1).isMainFile()) {
+            int pageNumber = files / filesPerPage;
+            int slotNumber = files % filesPerPage;
```

Comment:
I really don't like this math that we are doing here possible it would be easier if we would just store all project files in the Gui by making a helper class that would extend ItemStack but would also contain the ProjectFile instance that way we could easily determine if the item is project file and we could just filter pages until we find ItemStack that is instance of that helper class and we would also be able to get if that ItemStack is the main file